mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala

```diff
@@ -77,6 +77,11 @@ sealed trait Vector extends Serializable {
         // refer to {@link java.util.Arrays.equals} for hash algorithm
         val bits = java.lang.Double.doubleToLongBits(value)
         result = 31 * result + (bits ^ (bits >>> 32)).toInt
+        i += 1
+        // only scan the first 16 nonzeros
+        if (i > 16) {
+          return result
```

Comment:
`foreachActive` is usually on the critical path. This hashCode implementation is only for the cases when the vector is used as a map key, as in the Pyrolite SerDe. I added specialized version for DenseVector and SparseVector.